File: solutions/python/triangle/1/triangle.py
"""
A set of functions for checking whether a triangle 
is equilateral, isosceles or scalene.
"""

def triangle(sides):
    return sides[0] + sides[1] > sides[2] and sides[1] + sides[2] > sides[0] and sides[2] + sides[0] > sides[1]

def equilateral(sides):
    return triangle(sides) and sides[0] == sides[1] and sides[1] == sides[2]

def isosceles(sides):
    return triangle(sides) and (sides[0] == sides[1] or sides[1] == sides[2] or sides[2] == sides[0])

def scalene(sides):
    return triangle(sides) and not equilateral(sides) and not isosceles(sides)

File: solutions/fsharp/triangle/1/Triangle.fs
module Triangle

let positiveSides triangle = 
    let checkPositive x = if x > 0.0 then true else false
    List.map checkPositive triangle |> List.fold (fun a s -> a && s) true

let isTriangle (triangle: List<float>) =
    if triangle.[0] + triangle.[1] > triangle.[2] && triangle.[1] + triangle.[2] > triangle.[0] && triangle.[2] + triangle[0] > triangle.[1] then
        true
    else
        false

let equilateral triangle =
    if positiveSides triangle && isTriangle triangle then
        if triangle.[0] = triangle.[1] && triangle.[1] = triangle.[2] && triangle.[2] = triangle[0] then
            true
        else
            false
    else
        false

let isosceles triangle = 
    if positiveSides triangle && isTriangle triangle then
        if triangle.[0] = triangle.[1] || triangle.[1] = triangle.[2] || triangle.[2] = triangle[0] then
            true
        else
            false
    else
        false

let scalene triangle = 
    if positiveSides triangle && isTriangle triangle then
        if triangle.[0] <> triangle.[1] && triangle.[1] <> triangle.[2] && triangle.[2] <> triangle[0] then
            true
        else
            false
    else
        false
